File: java/day03/Solution.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Solution {
    public static void main(String[] args) {
        String inputFile = "src/input.txt";
        String regex = "\\((\\d+),(\\d+)\\)";
        Pattern pattern = Pattern.compile(regex);
        int mulResult = 0;

        try (BufferedReader reader = new BufferedReader(new FileReader(inputFile))) {
            String line;
            while ((line = reader.readLine()) != null) {
                Matcher matcher = pattern.matcher(line);
                while (matcher.find()) {
                    int num1 = Integer.parseInt(matcher.group(1));
                    int num2 = Integer.parseInt(matcher.group(2));
                    mulResult += num1 * num2;
                }
            }
        } catch (IOException e) {
            System.err.println("Error reading file: " + e.getMessage());
        }

        System.out.println("The sum of all multiplications is: " + mulResult);
    }
}

File: cpp/day03/solution.cpp
#include <fstream>
#include <iostream>
#include <regex>
#include <sstream>
#include <string>

using namespace std;

int main() {
  ifstream inputFile("input.txt");
  if (!inputFile.is_open()) {
    cerr << "Error opening the file!" << endl;
    return 1;
  }

  int mulResult = 0;
  string line;
  smatch matches;
  regex reg(R"(\((\d+),(\d+)\))");

  while (getline(inputFile, line)) {
    stringstream lineStream(line);

    while (lineStream >> line) {
      while (regex_search(line, matches, reg)) {
        mulResult += stoi(matches[1].str());
        mulResult += stoi(matches[2].str());
        line = matches.suffix().str();
      }
    }
  }
  inputFile.close();

  cout << "The sum of all multiplications is: " << mulResult << endl;

  return 0;
}
